lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -13,45 +13,58 @@
 
 #include <string>
 #include <vector>
+#include <set>
 
 #include "ShardTracker.h"
 
 namespace magma {
 
-int ShardTracker::add_ue() {
-  /*
-  * If there are no shards, add an entry for the shard
-  * representing a single UE at the new shard
-  */
-  if (ue_count_per_shard_.size() == 0) {
-    ue_count_per_shard_.push_back(1);
-    return 0;
+int ShardTracker::add_ue(std::string imsi) {
+  /**
+   * If there are no shards, add a single entry set
+   * to shards list containing the IMSI
+   * */
+  if (shards_.empty()) {
+    std::set<std::string> new_shard = {imsi};
+    shards_.push_back(new_shard);
   }
-  /*
-  * Iterate through all existing shards, if all are full
-  * create a new shard with quantity 1, otherwise increment
-  * the quantity of UEs in the latest shard
-  */
-  for (size_t shard_id = 0; shard_id < ue_count_per_shard_.size(); shard_id++) {
-    if (ue_count_per_shard_[shard_id] < max_shard_size_) {
-      ue_count_per_shard_[shard_id]++;
+
+  for (size_t shard_id = 0; shard_id < shards_.size(); shard_id++) {
+    /**
+     * If the UE is already in the shard, return the shard id. This check
+     * is meant to avoid multiple sessions for a UE being assigned duplicate
+     * entries.
+     * */
+    if (shards_[shard_id].find(imsi) != shards_[shard_id].end()) {
+      return shard_id;
+    }
+    /**
```

Comment:
this is my main concern on this design: it looks like the whole objective of this feature is to do groups of 100. 

But the question I still have is, what will you do with those groups? Are they going to be polled in parallel? How is this parallel work works? Does it uses multithread, multiprocess? Has there been any test done to see that parallel request to OVS do not clog ovs somewhere else?

What happen if you have 20 groups of 100? Will that mean you will poll one group each 10 seconds, which is equivalent to currently poll ovs every 200 seconds?

Reply:
It's easier for ovs/ryu to handle a bunch of small requests for subset of UEs than serve 1 huge one. We're going to try processing them in parallel(20 groups of 100 won't mean that we wait for one group reply to send a request for another one, we'll just attach a callback for once we recieve the stats to process the updates on sessiond)

Reply:
> @pshelar @koolzz I **strongly** recommend to have a design stating things like
> 
> * What is the issue
> * Where is the issue (is it in OVS, is it in pipelined): look for the functions that take longer on pipelined to understand where is this clogged.
> * Find or build a test case were you can reproduce this issue easily and quantify it. This should be used to compare once this is completed)
> * Purpose some solutions (multithread? multiprocessing?) and try to test them ONLY with pipelined (before polluting other systems)
> * If you are using shariding, do think possible sharding algorithms that may work, try to visualize how will they work (with pictures on design doc) before starting to build a shard structure.

those are valid points, ty for review:
I think we're confusing a few things I'm working on a design doc I'll share shortly but I think as this is just the skeleton for the sessiond->pipelined pull model which we all agreed on multiple times we shoud be fine merging this.

 - The issue is to scale up stats to support large number of UEs (returning all UEs * rules at once is becoming complicated)
 - Smart polling and moving to pull stats structure are 2 different efforts(even though connected smart polling is way more involved and design is not fully fleshed out yet). 
 - Sharding allows us to process stats for smaller groups of UEs, which is very convenient for datapath. 
 - We're not polluting sessiond, sessiond is the brains of the system, pipelined can't make any smart deicions because only sessiond know how much quota the UEs has. Down the line if we want to do any smart polling and group UEs with high traffic usage or break up GX/GY polling we need to do it from sessiond because we can't change the rules on pipelined side as they're installed by sessiond.

Reply:
right. the key point is 'scaling up'. Its not about the efficiency or bottleneck of existing stats reporting scheme. we can continue to improve it along with shard based stats reports.

Reply:
Have you guys tried to run parallel queries in pipelined to see what happens? Does it improve performance? Can you run 2 queries in parallel without performance impact, 20, 200?  What is the impact of a shard size of 100, what about 200 or 50? Answering those questions is basic to come up with a proper sharding

Why don't you start modifying pipelined to take parallel queries and do some ground testing to see if parallel queries are feasible?

Reply:
@veshkemburu will gather that on spirient. Although I'd prefer we land this PR to avoid the unnecessary confusion

Reply:
yeah I've landed it, we're sending sending a ovs req for specific shard (whenever we get a grpc req from pipelined) and whenever we get the result back we reply)

Reply:
What about duplicating the query you receive n times to see the performance?

Reply:
no tests yet unfortunately, only been back for a few days 😛